src/linkage.md

```diff
@@ -226,6 +226,24 @@ a statically linked binary on MSVC you would execute:
 RUSTFLAGS='-C target-feature=+crt-static' cargo build --target x86_64-pc-windows-msvc
 ```
 
+## Prohibited linkage scenarios
+
+No crate may be linked with [the `panic=abort` runtime][panic-runtime] if it has
+both of the following characteristics:
+
+* It contains a call to an `-unwind` foreign function or function pointer
```

Comment:
This includes "Rust" foreign function calls, right? Like a separate copy of the Rust runtime, built with the same compiler and hence ABI compatible.

In the other file you carefully define unwinding and non-unwinding ABIs; can that be reused here?

Reply:
Sorry, I'm not sure what you're asking with regard to runtimes. This is talking about the ABI strings `C-unwind`, `system-unwind`, etc; yes, the code being invoked may be Rust code.

What verbiage are you suggesting be reused? I don't think listing all the ABI strings seems necessary, but I can add a link to the `functions.md` section explaining unwinding vs non-unwinding ABIs, with the caveat that the Rust ABI itself doesn't prevent linking crates with different panic strategies.

Reply:
I'm saying that this needs to also include `extern "Rust" fn` foreign function calls, but the current wording does not seem to cover those functions.

Reply:
No, Rust calls are specifically excluded here; `fn` is entirely equivalent to `extern "Rust" fn`. So including `extern "Rust"` here would effectively mean "it's never legal to link `panic=abort` crates against `panic=unwind` crates," which is clearly not true.

... possibly I am still misunderstanding the question?

Reply:
Ah, I see. What makes "Rust" different from the other unwinding APIs?

IIRC the argument was that if we link in the abort runtime, Rust itself will never trigger unwinding, so if your program has no other "*-unwind" functions anywhere (meaning there's no foreign-language unwinding) then there can be no unwinding in general?

Reply:
Ugh, I thought I remembered how and why this rule was fashioned, but now that I'm thinking about it, I don't think I understand the problem it was intended to solve.

Part of what the "call to...foreign function" verbiage seems to be trying to capture is that the function call is actually linked to code from a language other than Rust, e.g. C++; because indeed, if that's the case, all bets are off about whether an unwind will occur or not. But I think that would apply equally to an opaque binary shared-object that happens to have been compiled from Rust code and linked against the `panic=unwind` runtime (so that in the final program, _both_ `panic=unwind` and `panic=abort` runtimes will be used).

If you compile with `panic=unwind` _and_ you use `extern "C"` (or other non-unwinding calls) everywhere that you call _into the other language runtime_ (be that C++, Rust, or something else), then you're safe: the `panic=unwind` crate has shims at each `extern "C"` boundary that will catch a panic and abort before it "escapes" into any stack frames compiled with `panic=abort`. But the `extern "C-unwind"` boundaries will have no such shims.

So I guess maybe `extern "Rust"` isn't an exception after all, if it's actually used to call _into_ an opaque shared object that might unwind; but is that even possible? Can you use an `extern "Rust"` call as an entrypoint into a "foreign" binary?

Reply:
Actually that may simply not be allowed? (Though that should probably be explicit.)

https://youtu.be/3pYHCGYJbw0?si=GbpPYif_OqI1UJrn

Reply:
The pure-Rust case is okay even if one object was built with `unwind=abort` and the other with `unwind=panic`, as long as the only panic _runtime_ linked is the `abort` runtime. This should be ensured even without Cargo just by duplicate-symbol linker errors. (And I'm pretty sure that compiling for `panic=abort` and then using the `panic=unwind` runtime is impossible with `rustc`, and if it were somehow managed, panicking would certainly be UB.)

So I think this additional restriction really does only matter in the case as described, where there's only one _Rust_ panic runtime, but a different language, such as C++, may introduce panics; in this case, you're safe _only if_ all the Rust stack frames that a C++ exception might touch are compiled with exception support. The phrasing in the PR ensures this because `extern "C"` will safely abort if a C++ exception comes through, _if_ that frame was compiled with `panic=unwind`.

Reply:
>  This should be ensured even without Cargo just by duplicate-symbol linker errors. 

How that?

> And I'm pretty sure that compiling for panic=abort and then using the panic=unwind runtime is impossible with rustc

Why is it impossible? What happens if I build one crate (type `lib`) with `-Cpanic=abort`, and then include it in a `bin` crate with `-Cpanic=unind`? Do we have a test for that.

Reply:
>> This should be ensured even without Cargo just by duplicate-symbol linker errors.
>
> How that?

Sorry, I was thinking duplicate-symbol link errors are guaranteed, but they're not. `rustc`/`lld` [does catch this error in](https://stackoverflow.com/questions/75993722/duplicate-symbol-rust-begin-unwind-when-mixing-rust-and-c) at least some cases. That said, if someone does actually link both the `abort` and the `panic` runtime, the result will be an ill-formed binary. I'm not sure whether or not that's possible with `rustc`, though; @bjorn3 wrote [here](https://users.rust-lang.org/t/how-rustc-manage-to-link-both-panic-abort-and-panic-unwind-on-the-same-binary/111518/2):

> Even though libstd declares a dependency on both the panic_abort and panic_unwind crates, rustc has special logic to only link one of the two depending on if code is compiled with `-Cpanic=abort` or `-Cpanic=unwind`.

I'm not sure where this special logic is, how it works, or how it's tested. If there's some way to get around it, I suppose it might be appropriate to document that here.

Reply:
Currently this seems to say that if you are mixing two Rust runtimes, one of which uses `-Cpanic=abort` and the other `-Cpanic=unwind`, and one calls the other via a `extern "Rust" fn`, then that is allowed. But this can't be correct. So it does seem to me like there is something that needs to be fixed here?

Reply:
No, omly the `abort` runtime is linked in the scenari described.

Reply:
You can freely mix code _compiled with_ different panic strategies, as long as the final binary only _links_ one runtime, and as long as the additional stipulation here about extern functions is followed.

Reply:
This line the comment is attached to is about calling code outside the current Rust runtime. It is reasonable to assume that it will give sufficient conditions for when it is okay to call code outside the current Rust runtime. Since only `-unwind` foreign functions are mentioned, then my reading of this text is that we would be *allowing* calls to Rust code outside the current Rust runtime via a "Rust" ABI fn. We don't want to allow such calls though, so this should be forbidden.
```suggestion
* It contains a call to an `-unwind` foreign function or function pointer, or a call to
  a "Rust" ABI function in another instance of the Rust runtime
```

Without this addition, the text seems just incomplete. I don't understand your last reply at all so, though it seems like we are increasingly talking past each other.

Reply:
If you use `cdylib` then you shouldn't use `extern "Rust"`. If you use `dylib` then it's not very different from statically linking two crates? You can't have two panic runtimes.

Reply:
> If you use cdylib then you shouldn't use extern "Rust".

Isn't that a thing people do, while also ensuring that it's the same Rust version on both sides?

Reply:
That's not guaranteed to work, but yes, in theory you can do it, and it may even work.

Reply:
Well, if people do it *and* not making sure panic agrees on both side, then it's not any different than using `extern "Rust"` but doesn't make sure ABI on both side match, e.g. different target feature. Also, this was unsound forever and not anything new due to `c_unwind`. `-Cpanic=abort` will make compiler assume `extern "Rust"` won't panic as early as Rust 1.23.

Reply:
> Also, this was unsound forever and not anything new due to c_unwind. -Cpanic=abort will make compiler assume extern "Rust" won't panic as early as Rust 1.23.

Absolutely, but now we're finally writing docs for all this so it seems odd to skip this one?

But okay, in the interest of unblocking this PR, maybe let's make this an issue then: https://github.com/rust-lang/reference/issues/1642.

Reply:
The text here certainly shouldn't be modified to add `extern "Rust"`, because the `ffi_unwind_calls` lint (in the paragraph below) works as described and does _not_ prevent using the `extern "Rust"` ABI.

Reply:
I would be okay with adding this sentence to the beginning of the "prohibited linkage scenarios" section, but I'm not 100% confident that it is accurate:

> It is never valid to link code against both [panic-runtimes][panic-runtime]; linkers will often reject attempts to do so by emitting a multiple-definitions error, but if a binary is successfully produced, any panic will result in undefined behavior.

In particular, I don't know what the right terminology is to distinguish these two types of linking:

* the same code is "linked against" both panic runtimes, in the sense that they both provide candidate definitions for the same symbol lookup at link-time (i.e., the situation I'm saying is never allowed, and which will tend to trigger multiple-definition errors)
* the situation @RalfJung described above: "a dylib...with -Cpanic=abort [loaded] into a Rust binary that was built with -Cpanic=unwind... The dylib statically contains a Rust runtime, the main binary statically contains another independent Rust runtime."

Reply:
@RalfJung I've rewritten this section to make a differentiation that I believe is correct (though I'd prefer if someone else could confirm): the undefined behavior caused by unwinding into a Rust binary that doesn't have unwinding support, versus the protection provided at link-time by `rustc`. In particular, you're correct that _if_ a user managed to use an `extern "Rust"` function call to invoke an entrypoint in a static or shared library, and that call unwinds, then there's undefined behavior; but the link-error designed to protect against this operates under the assumption that `extern "Rust"` will not be used this way, as discussed above.

Additionally, since I'm not sure why this prohibition would apply to an `exe` with no "foreign" (i.e. precompiled) code, I wrote that it applies "even when linking a static or dynamic library that only includes Rust code".

Reply:
I thought we had reached a point where the text was fine, if we keep changing it this will never terminate...

> I don't remember how or where this safeguard in rustc is implemented. Does it apply even if you're building an exe and only Rust code is involved?

Yes. Rust code built with -Cpanic=abort generates different MIR than Rust code built with -Cpanic=unwind, and mixing both in the same binary can unsound, hence the stated requirements.